pkgs/by-name/ni/nix-required-mounts/main.py

```diff
@@ -0,0 +1,121 @@
+import glob
+import json
+import subprocess
+import textwrap
+from argparse import ArgumentParser
+from itertools import chain
+from pathlib import Path
+from sys import stderr
+from typing import List
+
+CONFIG = {
+    "nixExe": "nix",
+    "allowedPatterns": {},
+}
+
+parser = ArgumentParser("pre-build-hook")
+parser.add_argument("derivation_path")
+parser.add_argument("sandbox_path", nargs="?")
+parser.add_argument(
+    "--issue-command",
+    choices=("always", "conditional", "never"),
+    default="conditional",
+    help="Whether to print extra-sandbox-paths",
+)
+parser.add_argument(
+    "--issue-stop",
+    choices=("always", "conditional", "never"),
+    default="conditional",
+    help="Whether to print the final empty line",
+)
+
+
+def symlink_parents(p: Path) -> List[Path]:
+    out = []
+    while p.is_symlink() and p not in out:
+        p = p.readlink()
+        out.append(p)
+    return out
+
+
+def entrypoint():
+    if __name__ != "__main__":
+        return
+
+    args = parser.parse_args()
+    drv_path = args.derivation_path
+
+    if not Path(drv_path).exists():
+        print(
+            f"[E] {drv_path} doesn't exist."
+            " This may happen with the remote builds."
+            " Exiting the hook",
+            file=stderr,
+        )
+
+    proc = subprocess.run(
+        [
+            CONFIG["nixExe"],
+            "show-derivation",
+            drv_path,
+        ],
+        capture_output=True,
+    )
+    try:
+        drv = json.loads(proc.stdout)
+    except json.JSONDecodeError:
+        print(
+            "[E] Couldn't parse the output of"
+            "`nix show-derivation`"
+            f". Expected JSON, observed: {proc.stdout}",
+            file=stderr,
+        )
+        print(
+            textwrap.indent(proc.stdout.decode("utf8"), prefix=" " * 4),
+            file=stderr,
+        )
+        print("[I] Exiting the nix-required-binds hook", file=stderr)
+        return
+    [canon_drv_path] = drv.keys()
+
+    allowed_patterns = CONFIG["allowedPatterns"]
+    known_features = set(
+        chain.from_iterable(
+            pattern["onFeatures"] for pattern in allowed_patterns.values()
+        )
+    )
+
+    drv_env = drv[canon_drv_path].get("env", {})
+    features = drv_env.get("requiredSystemFeatures", [])
+    if isinstance(features, str):
+        features = features.split()
```

Comment:
```suggestion
    features = drv_env.get("requiredSystemFeatures", "").split()
```

Reply:
Are `env.*` values always going to be a string? Whatever happens when one uses structured attrs?

Reply:
IIUC they should be exactly equivalent: 
* if `"requiredSystemFeatures"` exists in `drv_env`, then `drv_env.get("requiredSystemFeatures", "").split() => drv_env["requiredSystemFeatures"].split()`
* if `"requiredSystemFeatures"` does not exist in `drv_env`, then `drv_env.get("requiredSystemFeatures", "").split() => "".split() => []`.

Reply:
What I meant is, I thought structuredAttrs would prevent requiredSystemFeatures the list from collapsing into a string. But I didn't check if the list really goes into the env.*

Reply:
Maybe refer to the code that nix uses to resolve it?

https://github.com/NixOS/nix/blob/201c115c3e242e3db60a2cd1583223d4e196d6e5/src/libstore/parsed-derivations.cc#L64-L99

If `structuredAttrs`, then it is a array of strings in the "__json" json serialization.
Else it is a whitespace-separated string in the environment.

Reply:
Thanks @Kiskae! Rewrote the bit in the last commit